src/staging/validation.py

```diff
@@ -631,3 +631,48 @@ def validate_cora_df(df: pd.DataFrame) -> pd.DataFrame:
 
     except ValueError as ve:
         raise ValueError("cora status mapper validation failed: " + str(ve))
+
+
+@exception_wrap
+def update_ref_list(full_df: pd.DataFrame, ref_list_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Update long form references that should be on the reference list.
+
+    For the first year (processing 2022 data) only, several references
+    should have been designated on the "reference list", ie, should have been
+    assigned cellnumber = 817, but were wrongly assigned a different cellnumber.
+
+    Args:
+        full_df (pd.DataFrame): The full_responses dataframe
+        ref_list_df (pd.DataFrame): The mapper containing updates for the cellnumber
+    Returns:
+        df (pd.DataFrame): with cellnumber and selectiontype cols updated.
+    """
+    ref_list_filtered = ref_list_df.loc[
+        (ref_list_df.formtype == 1) & (ref_list_df.cellnumber != 817)
+    ]
+    df = pd.merge(
+        full_df, 
+        ref_list_filtered[["reference", "cellnumber"]], 
+        how="left", 
+        on="reference", 
+        suffixes=("", "_new"), 
+        indicator=True
+    )
+    # check no items in the reference list mapper are missing from the full responses
+    missing_refs = df.loc[df["_merge"] == "right_only"]
+    if not missing_refs.empty:
+        msg = (
+        "The following references in the reference list mapper are not in the data:"
+        )
+        raise ValueError(msg + str(missing_ref.reference.unique()))
```

Comment:
I like adding this validation - I wouldn't have thought to do it!

I think you should keep it like this as it's clearer, but just as an extra you can validate pandas merges are many to one with a `validate` argument in `pd.merge`. I don't think though it would check if there are "un matched" values.

Reply:
Nice to know! I'll look that up